internal/workflow/localact/save_ingest_tasks.go

```diff
@@ -0,0 +1,81 @@
+package localact
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"io"
+
+	"github.com/google/uuid"
+
+	"github.com/artefactual-sdps/enduro/internal/datatypes"
+	"github.com/artefactual-sdps/enduro/internal/enums"
+	"github.com/artefactual-sdps/enduro/internal/ingest"
+	"github.com/artefactual-sdps/enduro/internal/package_"
+)
+
+type SaveIngestTasksActivityParams struct {
+	// PkgSvc is a package service instance.
+	PkgSvc package_.Service
+
+	// RNG is a random number generator source.
+	RNG io.Reader
+
+	// PreservationActionID is the primary key of the parent PreservationAction.
+	PreservationActionID uint
+
+	// Tasks is a list of preprocessing tasks to save as PreservationTasks.
+	Tasks []ingest.Task
+}
+
+type SaveIngestTasksActivityResult struct {
+	// Count is the number of saved PreservationTasks.
+	Count int
+}
+
+func SaveIngestTasksActivity(
+	ctx context.Context,
+	params SaveIngestTasksActivityParams,
+) (*SaveIngestTasksActivityResult, error) {
+	var res SaveIngestTasksActivityResult
+	for _, t := range params.Tasks {
+		pt := ingestTaskToPresTask(t)
+		pt.PreservationActionID = params.PreservationActionID
+
+		u, err := uuid.NewRandomFromReader(params.RNG)
+		if err != nil {
+			return &res, fmt.Errorf("SaveIngestTasksActivity: generate UUID: %v", err)
+		}
+		pt.TaskID = u.String()
+
+		if err := params.PkgSvc.CreatePreservationTask(ctx, &pt); err != nil {
+			return &res, fmt.Errorf("SaveIngestTasksActivity: %v", err)
+		}
+		res.Count++
+	}
+
+	return &res, nil
+}
+
+func ingestTaskToPresTask(t ingest.Task) datatypes.PreservationTask {
+	taskOutcomeToPresTaskStatus := map[enums.IngestTaskOutcome]enums.PreservationTaskStatus{
+		enums.IngestTaskOutcomeUnspecified:       enums.PreservationTaskStatusUnspecified,
+		enums.IngestTaskOutcomeSuccess:           enums.PreservationTaskStatusDone,
+		enums.IngestTaskOutcomeSystemFailure:     enums.PreservationTaskStatusError,
+		enums.IngestTaskOutcomeValidationFailure: enums.PreservationTaskStatusError,
+	}
+
+	status, found := taskOutcomeToPresTaskStatus[t.Outcome]
+	if !found {
+		status = enums.PreservationTaskStatusUnspecified
+	}
+
+	return datatypes.PreservationTask{
+		TaskID:      uuid.New().String(),
+		Name:        t.Name,
+		Status:      status,
+		StartedAt:   sql.NullTime{Time: t.StartedAt, Valid: true},
+		CompletedAt: sql.NullTime{Time: t.CompletedAt, Valid: true},
```

Comment:
Should we check they are not zero dates?

Reply:
Yes, good idea.  I've added a helper function to convert `time.Time` to `db.NullTime` properly, including zero times.

We've implemented a complementary conversion `db.FormatOptionalTime` in the database package, which makes me realize we shouldn't be using an implementation specific type (`db.NullTime`) outside of the DB implementation layer.  I will change the types of the `StartedAt` and `CompletedAt` fields in a future PR.